File: blue/src/package_neon_multi_node_blue/ssh_config.py
"""Package-owned atomic SSH aliases; colors-compute owns keys and topology."""

from __future__ import annotations

import os
import re
from pathlib import Path

from colors_compute.contract import expand

from . import topology


def host_alias(opts: dict) -> str:
    """The profile, unchanged. Standard §2: the profile already keys remote
    state, which is what makes it unique enough to name a host by. It reaches
    the PostgreSQL compute host."""
    return opts.get("profile") or "neon-multi-node"


def identity_file(opts: dict) -> str:
    """`~/.ssh/<profile>`, written with a literal tilde rather than an expanded
    home directory. OpenSSH expands it, and leaving it unexpanded is what keeps
    the rendered block identical on every workstation."""
    return f"~/.ssh/{host_alias(opts)}"


def aliases(opts):
    return [host_alias(opts), *[host_alias(opts) + '-' + n['node_id'] for n in expand(topology.topology(opts))]]


def machine_alias(opts, host):
    return host_alias(opts) + '-' + host['node_id']


def config_path() -> Path:
    home = os.environ.get("HOME")
    return (Path(home) if home else Path.home()) / ".ssh" / "config"


def begin_marker(alias: str) -> str:
    return f"# BEGIN {alias} ANSIBLE MANAGED BLOCK"


def end_marker(alias: str) -> str:
    return f"# END {alias} ANSIBLE MANAGED BLOCK"


def owned_markers(alias: str) -> dict:
    return {"begin": {begin_marker(alias)}, "end": {end_marker(alias)}}


def host_patterns(line: str) -> list[str] | None:
    """The patterns a `Host` line declares, or None when the line is not one."""
    match = re.fullmatch(r"(?i)\s*Host\s+(.*?)\s*", str(line))
    if not match:
        return None
    return [p for p in re.split(r"\s+", match.group(1)) if p.strip()]


def foreign_stanza_line(lines: list, alias: str, marker_alias: str | None = None) -> int | None:
    """The 1-based line number of a `Host <alias>` stanza that this package
    did not write, or None. Lines between our own markers are ours and are
    skipped.

    `alias` is the stanza being searched for; `marker_alias` names the managed
    block, and the two are not the same thing: this deployment writes ONE
    block, marked with the profile, containing a stanza for the profile and
    for every machine."""
    markers = owned_markers(alias if marker_alias is None else marker_alias)
    inside = False
    for n, line in enumerate(lines, start=1):
        trimmed = str(line).strip()
        if trimmed in markers["begin"]:
            inside = True
        elif trimmed in markers["end"]:
            inside = False
        elif not inside and alias in (host_patterns(line) or []):
            return n
    return None


def leading_option_line(lines: list) -> int | None:
    """The 1-based line number of an option standing above the first `Host`
    or `Match` line, or None. Such an option is global; the block is written
    with `insertbefore: BOF`, so it would capture that option into this
    deployment's stanza, silently narrowing a global setting to one host."""
    for n, line in enumerate(lines, start=1):
        trimmed = str(line).strip()
        if not trimmed or trimmed.startswith("#"):
            continue
        if re.fullmatch(r"(?i)\s*(Host|Match)\s+.*", str(line)):
            return None
        return n
    return None


def adopt_error(opts: dict) -> str | None:
    """The standard's never-adopt rule (§5), checked for every alias this
    deployment claims."""
    file = config_path()
    if not file.is_file():
        return None
    lines = file.read_text().splitlines()
    marker = host_alias(opts)
    for alias in aliases(opts):
        n = foreign_stanza_line(lines, alias, marker)
        if n is not None:
            return (f"refusing to manage {file}: it already declares "
                    f"`Host {alias}` at line {n}"
                    " outside this package's managed block. Remove or rename that "
                    "stanza if it is stale, or change `profile` if it belongs to "
                    "something else; this package will not overwrite it.")
    return None


def placement_error(_opts: dict) -> str | None:
    file = config_path()
    if not file.is_file():
        return None
    n = leading_option_line(file.read_text().splitlines())
    if n is None:
        return None
    return (f"refusing to manage {file}: line {n}"
            " sets an option above the first `Host` line, so it applies to "
            "every host. This package inserts its block at the top of the "
            "file, which would capture that option into one stanza. Move "
            "those global options below the managed block, or into an "
            "explicit `Host *` stanza at the end of the file, and retry.")


def preflight(opts: dict) -> dict:
    """Run the local checks. Real create only: build and dry-run must not read
    `~/.ssh/config` at all (§6)."""
    error = adopt_error(opts) or placement_error(opts)
    if error:
        return {**opts, "blue/exit": 1, "blue/err": error}
    return opts

File: red/src/ssh-config.ts
// Package-owned SSH aliases and never-adopt checks for the joined five-node inventory.
import { existsSync, readFileSync, statSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import type { Opts } from "red/workflow";
import {expand} from "colors-compute-red";
import * as topology from "./topology.ts";

// The profile, unchanged. Standard §2: the profile already keys remote state,
// which is what makes it unique enough to name a host by. It reaches the compute
// host used for PostgreSQL access.
export function hostAlias(opts: Opts): string {
  return String(opts.profile ?? "neon-multi-node");
}

// `~/.ssh/<profile>`, written with a literal tilde rather than an expanded
// home directory. OpenSSH expands it, and leaving it unexpanded is what keeps
// the rendered block identical on every workstation.
export function identityFile(opts: Opts): string {
  return `~/.ssh/${hostAlias(opts)}`;
}

// Entry alias plus stable profile-node_id alias for each compute-library node.
export function aliases(opts:Opts):string[]{return [hostAlias(opts),...expand(topology.topology(opts)).map(n=>hostAlias(opts)+'-'+n.node_id)];}
export function machineAlias(opts:Opts,host:topology.Host):string{return hostAlias(opts)+'-'+host.node_id;}

export function configPath(): string {
  return join(process.env.HOME ?? homedir(), ".ssh", "config");
}

export function beginMarker(alias: string): string {
  return `# BEGIN ${alias} ANSIBLE MANAGED BLOCK`;
}

export function endMarker(alias: string): string {
  return `# END ${alias} ANSIBLE MANAGED BLOCK`;
}

export function ownedMarkers(alias: string): { begin: Set<string>; end: Set<string> } {
  return { begin: new Set([beginMarker(alias)]), end: new Set([endMarker(alias)]) };
}

// The patterns a `Host` line declares, or undefined when the line is not one.
export function hostPatterns(line: string): string[] | undefined {
  const match = /^\s*Host\s+(.*?)\s*$/i.exec(line);
  if (!match) return undefined;
  return match[1]!.split(/\s+/).filter((pattern) => pattern.length > 0);
}

// The 1-based line number of a `Host <alias>` stanza that this package did not
// write, or undefined. Lines between our own markers are ours and are skipped.
//
// `alias` is the stanza being searched for; `markerAlias` names the managed
// block, and the two are not the same thing: this deployment writes ONE block,
// marked with the profile, containing a stanza for the profile and for every
// machine.
export function foreignStanzaLine(lines: string[], alias: string, markerAlias = alias): number | undefined {
  const { begin, end } = ownedMarkers(markerAlias);
  let inside = false;
  for (let i = 0; i < lines.length; i += 1) {
    const line = lines[i]!;
    if (begin.has(line.trim())) inside = true;
    else if (end.has(line.trim())) inside = false;
    else if (!inside && (hostPatterns(line) ?? []).includes(alias)) return i + 1;
  }
  return undefined;
}

// The 1-based line number of an option standing above the first `Host` or
// `Match` line, or undefined. Such an option is global; the block is written
// with `insertbefore: BOF`, so it would capture that option into this
// deployment's stanza, silently narrowing a global setting to one host.
export function leadingOptionLine(lines: string[]): number | undefined {
  for (let i = 0; i < lines.length; i += 1) {
    const trimmed = String(lines[i]).trim();
    if (trimmed === "" || trimmed.startsWith("#")) continue;
    if (/^\s*(Host|Match)\s+.*/i.test(lines[i]!)) return undefined;
    return i + 1;
  }
  return undefined;
}

function configFileLines(): string[] | undefined {
  const path = configPath();
  if (!existsSync(path) || !statSync(path).isFile()) return undefined;
  return readFileSync(path, "utf8").split("\n");
}

// The standard's never-adopt rule (§5), checked for every alias this
// deployment claims.
export function adoptError(opts: Opts): string | undefined {
  const lines = configFileLines();
  if (!lines) return undefined;
  const marker = hostAlias(opts);
  for (const alias of aliases(opts)) {
    const n = foreignStanzaLine(lines, alias, marker);
    if (n === undefined) continue;
    return `refusing to manage ${configPath()}: it already declares ` +
      `\`Host ${alias}\` at line ${n}` +
      " outside this package's managed block. Remove or rename that " +
      "stanza if it is stale, or change `profile` if it belongs to " +
      "something else; this package will not overwrite it.";
  }
  return undefined;
}

export function placementError(_opts: Opts): string | undefined {
  const lines = configFileLines();
  if (!lines) return undefined;
  const n = leadingOptionLine(lines);
  if (n === undefined) return undefined;
  return `refusing to manage ${configPath()}: line ${n}` +
    " sets an option above the first `Host` line, so it applies to " +
    "every host. This package inserts its block at the top of the " +
    "file, which would capture that option into one stanza. Move " +
    "those global options below the managed block, or into an " +
    "explicit `Host *` stanza at the end of the file, and retry.";
}

export interface PreflightChecks {
  adoptError: (opts: Opts) => string | undefined;
  placementError: (opts: Opts) => string | undefined;
}

// Run the local checks. Real create only: build and dry-run must not read
// `~/.ssh/config` at all (§6). The checks are injectable so tests can cover
// the refusal without a doctored home directory.
export function preflight(
  opts: Opts,
  checks: PreflightChecks = { adoptError, placementError },
): Opts {
  const error = checks.adoptError(opts) ?? checks.placementError(opts);
  return error ? { ...opts, "red/exit": 1, "red/err": error } : opts;
}
